src/frontend/src/binder/expr/mod.rs

```diff
@@ -536,6 +596,26 @@ impl Binder {
             .transpose()?;
 
         let mut constant_lookup_inputs = Vec::new();
+        let mut constant_case_when_eval_inputs = Vec::new();
+
+        let constant_case_when_flag = self.check_constant_case_when_optimization(
+            conditions.clone(),
+            results_expr.clone(),
+            operand.clone(),
+            else_result_expr.clone(),
+            &mut constant_case_when_eval_inputs,
+        );
+
+        if constant_case_when_flag {
+            // Here we reuse the `ConstantLookup` as the `FunctionCall`
+            // to avoid creating a dummy `ConstCaseWhenEval` expression type
+            // since we do not need to go through backend
+            return Ok(FunctionCall::new(
+                ExprType::ConstantLookup,
+                constant_case_when_eval_inputs,
+            )?
+            .into());
```

Comment:
Can we just return `constant_case_when_eval_inputs[0]` here? Then we don't even need to use a rule to convert it later.

Reply:
Then what type should we return, I doubt FunctionCall would work 🤔, also Literal may not be the appropriate one.

Reply:
This method returns an `ExprImpl`, an enum that could be any expression type.

Reply:
Got it!
Then I think return a bound expression should work, and I'll remove the relevant rules. 😄